### PR TITLE
fix(web): correctly compute localized downtime's datepickers

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

Fix an issue where datetime of downtimes weren't correctly computed, when not using the english localization :
- the ending datepicker was not set properly and displayed unconsistent date
- adding a downtime on two days, did not actualize the starting date of the datepicker
- setting an end time before the start date was possible and resulted on useless downtime (not processed by centengine)

**Fixes** MON-3935 & https://github.com/centreon/centreon/pull/7606 from @UrBnW  + https://github.com/centreon/centreon/issues/5963 + https://github.com/centreon/centreon/issues/7337 + https://github.com/centreon/centreon/issues/8266

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [x] 19.04.x
- [x] 19.10.x
- [x] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

- Choose for example the french language.
- Select a default duration on the Administration > parameters > monitoring + duration
- Set a downtime from the monitoring > status > downtime + add downtime
    check that the start and end date and time are set properly.
    check that consistent values are displayed when changing the datepicker and timepicker values.
    check that the datepicker and timepicker are still consistent with the choosen language.

- Do the same from all the other available downtime form in the UI eg :
    status > monitoring > services + choose a resource + set a downtime from the toolbar
    status > monitoring > host + choose a resource + set a downtime from the toolbar
    ...

- Check that no errors are displayed in the logs and that the moment deprecated notice is corrected :
image

## Checklist

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/include/monitoring/downtime/AddDowntime.php [hunk 1]
```diff
@@ -287,6 +278,17 @@
     );
 
     $defaultDuration = 7200;
+    $defaultScale = 's';
+    if (isset($centreon->optGen['monitoring_dwt_duration']) &&
+        $centreon->optGen['monitoring_dwt_duration']
+    ) {
```
Comment (sc979) [inline www/include/monitoring/downtime/AddDowntime.php]:
```suggestion
    if (
        isset($centreon->optGen['monitoring_dwt_duration'])
        && $centreon->optGen['monitoring_dwt_duration']
    ) {
```

File: www/include/monitoring/downtime/AddDowntime.php [hunk 2]
```diff
@@ -287,6 +278,17 @@
...
+        $defaultDuration = $centreon->optGen['monitoring_dwt_duration'];
+        if (isset($centreon->optGen['monitoring_dwt_duration_scale']) &&
+            $centreon->optGen['monitoring_dwt_duration_scale']
+        ) {
+            $defaultScale = $centreon->optGen['monitoring_dwt_duration_scale'];
```
Comment (sc979) [inline www/include/monitoring/downtime/AddDowntime.php]:
```suggestion
        if (
            isset($centreon->optGen['monitoring_dwt_duration_scale'])
            && $centreon->optGen['monitoring_dwt_duration_scale']
        ) {
            $defaultScale = $centreon->optGen['monitoring_dwt_duration_scale'];
```

File: www/include/monitoring/external_cmd/popup/massive_downtime.php [hunk 3]
```diff
@@ -149,7 +145,20 @@
         'disabled' => 'true'
     )
 );
+//setting default values
 $defaultDuration = 7200;
+$defaultScale = 's';
+//overriding the default duration and scale by the user's value from the administration fields
+if (isset($centreon->optGen['monitoring_dwt_duration']) &&
+    $centreon->optGen['monitoring_dwt_duration']
+) {
```
Comment (sc979) [inline www/include/monitoring/external_cmd/popup/massive_downtime.php]:
```suggestion
if (
    isset($centreon->optGen['monitoring_dwt_duration'])
    && $centreon->optGen['monitoring_dwt_duration']
) {
```

File: www/include/monitoring/external_cmd/popup/massive_downtime.php [hunk 4]
```diff
@@ -149,7 +145,20 @@
...
+    $defaultDuration = $centreon->optGen['monitoring_dwt_duration'];
+    if (isset($centreon->optGen['monitoring_dwt_duration_scale']) &&
+        $centreon->optGen['monitoring_dwt_duration_scale']
+    ) {
```
Comment (sc979) [inline www/include/monitoring/external_cmd/popup/massive_downtime.php]:
```suggestion
    if (
        isset($centreon->optGen['monitoring_dwt_duration_scale'])
        && $centreon->optGen['monitoring_dwt_duration_scale']
    ) {
```

File: www/include/monitoring/external_cmd/popup/templates/massive_downtime.ihtml [hunk 5]
```diff
@@ -61,10 +61,12 @@
 </form>
 {literal}
 <script type="text/javascript">
-    /* initializing datepicker and the alternative format field */
-    initDatepicker("datepicker", "mm/dd/yy", "0");
-
+    jQuery(".timepicker").each(function () {
+        $(this).val(moment().tz(localStorage.getItem('realTimezone') ? localStorage.getItem('realTimezone') : moment.tz.guess()).format("HH:mm"));
+    });
```
Comment (sc979) [inline www/include/monitoring/external_cmd/popup/templates/massive_downtime.ihtml]:
```suggestion
    jQuery(".timepicker").each(function () {
        $(this).val(moment().tz(localStorage.getItem('realTimezone')
            ? localStorage.getItem('realTimezone')
            : moment.tz.guess()).format("HH:mm")
        );
    });
```

File: www/include/common/javascript/datepicker/localizedDatepicker.js [hunk 6]
```diff
@@ -62,19 +65,22 @@ function initDatepicker(className, altFormat, defaultDate, idName, timestampToSe
             var altName = jQuery(this).attr("name");
             if (typeof(altName) != "undefined") {
                 var alternativeField = "input[name=alternativeDate" + altName[0].toUpperCase() + altName.slice(1) + "]";
-                var value = $(this) && $(this).val() ? $(this).val() : defaultDate;
+                // $(this).val(), if exists, is a GMT YYYY-MM-DDTHH:mm:ss timestamp, for example with PHP : gmdate("Y-m-d\TH:i:s")
+                var value = $(this) && $(this).val() ? moment($(this).val(), "DD-MM-YYYY").tz(timezone) : defaultDate;
```
Comment (UrBnW) [inline www/include/common/javascript/datepicker/localizedDatepicker.js]: IMO `, "DD-MM-YYYY"` you added is incorrect / not needed.
Look at the comment above (and also line 40) saying that a GMT `YYYY-MM-DDTHH:mm:ss` timestamp is expected.

Comment (sc979) [reply www/include/common/javascript/datepicker/localizedDatepicker.js]: Hi @UrBnW ,
The format was added to resolve the deprecated moment notice when submitting the form :
image

File: www/include/monitoring/external_cmd/popup/massive_downtime.php [hunk 7]
```diff
@@ -107,14 +102,19 @@
     )
 );
 
+$userDownTimeDuration = $_SESSION['centreon']->optGen['monitoring_dwt_duration'];
+$defaultStartTime = date('G:i');
+$defaultEndTime = date('G:i', time() + $userDownTimeDuration);
```
[the remark this answers is not in the document]
Comment (sc979) [reply www/include/monitoring/external_cmd/popup/massive_downtime.php]: Indeed, the timepicker was not instanciated in the downtime form in the popup from monitoring > status > service or host + select a resource and add a downtime from the toolbar : 
image

[the remark this answers is not in the document]
Comment (sc979) [reply www/include/monitoring/external_cmd/popup/massive_downtime.php]: For the widgets, the PR will be made on their repo

File: www/include/common/javascript/datepicker/localizedDatepicker.js (hunk 6 above)
Comment (sc979) [reply www/include/common/javascript/datepicker/localizedDatepicker.js]: The modification is related to this issue : https://github.com/moment/moment/issues/1407

File: www/include/monitoring/external_cmd/popup/massive_downtime.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (UrBnW) [reply www/include/monitoring/external_cmd/popup/massive_downtime.php]: I do not face your issue.
IMO you should not need to set time at backend.
Time is set at frontend, in `AddDowntime.ihtml` / `massive_downtime.ihtml`, thanks to Moments.js and jQuery.
Strange then 😕

[the remark this answers is not in the document]
Comment (sc979) [reply www/include/monitoring/external_cmd/popup/massive_downtime.php]: It resulted from this suppression : https://github.com/centreon/centreon/pull/7606/files#diff-d73252cdaa3d27a76bf712a18addbabcL324
Then the backend never set the timepicker start and end date and the field are empty.
The front end part do set this value only when you choose at least one value (start or end)

[the remark this answers is not in the document]
Comment (UrBnW) [reply www/include/monitoring/external_cmd/popup/massive_downtime.php]: > The front end part do set this value only when you choose at least one value (start or end)

I'm not sure, look at the code added by this PR into `AddDowntime.ihtml` / `massive_downtime.ihtml` :

```
    jQuery(".timepicker").each(function () {
        $(this).val(moment().tz(localStorage.getItem('realTimezone')
            ? localStorage.getItem('realTimezone')
            : moment.tz.guess()).format("HH:mm")
        );
```

File: www/include/common/javascript/datepicker/localizedDatepicker.js (hunk 6 above)
Comment (UrBnW) [reply www/include/common/javascript/datepicker/localizedDatepicker.js]: Do you know where this value was provided from ?
I'm pretty sure not from downtimes, as downtime dates are not computed by the backend anymore, but by the frontend / Moment.js.
Perhaps we should locate the culprit(s) and modify the format they use to feed the datepicker, to a proper GMT `YYYY-MM-DDTHH:mm:ss` one ?

File: www/include/monitoring/external_cmd/popup/massive_downtime.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (UrBnW) [reply www/include/monitoring/external_cmd/popup/massive_downtime.php]: A browser cache issue perhaps which prevents this piece of code from properly running ? :)

File: www/include/common/javascript/datepicker/localizedDatepicker.js (hunk 6 above)
Comment (UrBnW) [reply www/include/common/javascript/datepicker/localizedDatepicker.js]: You should then need to merge https://github.com/centreon/centreon/pull/8289 before :)

Comment (sc979) [reply www/include/common/javascript/datepicker/localizedDatepicker.js]: it's already merger on the master. specific PR are required to add theses modifications on the other branches

**Comment by sc979** [comment]
some behavior weren't corrected
The PR https://github.com/centreon/centreon/pull/8486, take in account all feedbacks.
Thanks again @UrBnW for your time

**Pull request closed without merging**